File: kernel/src/test/java/com/sos/scheduler/engine/kernel/plugin/PlugInAdapterTest.java
package com.sos.scheduler.engine.kernel.plugin;

import com.sos.scheduler.engine.kernel.Platform;
import com.sos.scheduler.engine.kernel.PlatformMock;
import org.junit.Test;
import static com.sos.scheduler.engine.kernel.util.XmlUtils.loadXml;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.*;


public class PlugInAdapterTest {
    private final Platform platform = PlatformMock.newInstance();
    private final PluginAdapter plugInAdapter = new PluginAdapter(new PlugInMock(), "plugInMock", platform.log());
    private final PluginAdapter errorPlugInAdapter = new PluginAdapter(new ErrorPlugInMock(), "errorPlugInMock", platform.log());


    @Test public void testActivate() {
        plugInAdapter.activate();
    }


    @Test(expected=RuntimeException.class) public void testActivate_error() {
        errorPlugInAdapter.activate();
    }


    @Test public void testClose() {
        plugInAdapter.close();
    }


    @Test(expected=RuntimeException.class) public void testClose_error() {
        errorPlugInAdapter.close();
    }


    @Test public void testGetXmState() {
        loadXml(plugInAdapter.getXmlState());
        errorPlugInAdapter.getXmlState();
    }

    
    @Test public void testTryActivate() {
        plugInAdapter.tryActivate();
        errorPlugInAdapter.tryActivate();
    }


    @Test public void testTryClose() {
        plugInAdapter.tryClose();
        errorPlugInAdapter.tryClose();
    }

    @Test public void testGetXmlState() {
        loadXml(plugInAdapter.getXmlState());
        loadXml(errorPlugInAdapter.getXmlState());
    }

    @Test public void testToString() {
        assertThat(plugInAdapter.toString(), startsWith("Plugin "));
    }
}

File: engine-kernel/src/test/scala/com/sos/scheduler/engine/kernel/plugin/PluginAdapterTest.scala
package com.sos.scheduler.engine.kernel.plugin

import com.google.inject.Guice._
import com.google.inject.Injector
import com.sos.scheduler.engine.common.xml.CppXmlUtils.loadXml
import com.sos.scheduler.engine.kernel.plugin.PluginAdapterTest._
import org.junit.runner.RunWith
import org.scalatest.FunSuite
import org.scalatest.Matchers._
import org.scalatest.junit.JUnitRunner

@RunWith(classOf[JUnitRunner])
final class PluginAdapterTest extends FunSuite {
  private val pluginAdapter = new PluginAdapter(pluginConfiguration)
  private val errorPluginInstanceAdapter = new PluginAdapter(errorPluginConfiguration)
  private val injector: Injector = createInjector()

  test("initialize") {
    pluginAdapter.initialize(injector)
  }

  test("prepare") {
    pluginAdapter should not be 'prepared
    pluginAdapter.prepare()
    pluginAdapter shouldBe 'prepared
  }

  test("activate") {
    pluginAdapter should not be 'active
    pluginAdapter.activate()
    pluginAdapter shouldBe 'active
  }

  test("activate error") {
    errorPluginInstanceAdapter.initialize(injector)
    intercept[RuntimeException] { errorPluginInstanceAdapter.activate() }
  }

  test("tryActivate") {
    pluginAdapter shouldBe 'active
    pluginAdapter.tryActivate()
    pluginAdapter shouldBe 'active
    errorPluginInstanceAdapter.tryActivate()
    errorPluginInstanceAdapter should not be 'active
  }

  test("tryClose") {
    pluginAdapter shouldBe 'active
    pluginAdapter.tryClose()
    pluginAdapter should not be 'active
    errorPluginInstanceAdapter should not be 'active
    errorPluginInstanceAdapter.tryClose()
    errorPluginInstanceAdapter should not be 'active
  }

  test("xmlState") {
    loadXml(pluginAdapter.xmlState)
    loadXml(errorPluginInstanceAdapter.xmlState)
  }

  test("toString") {
    pluginAdapter.toString should startWith ("Plugin ")
  }
}

private object PluginAdapterTest {
  private val pluginConfiguration = new PluginConfiguration(classOf[MockPlugin].getName, None)
  private val errorPluginConfiguration = new PluginConfiguration(classOf[ErrorMockPlugin].getName, None)
}
